From d366dd91006a430540e1a1a695e421dc5902dd0b Mon Sep 17 00:00:00 2001
From: S8Cloud <github@live.hk>
Date: Wed, 7 Nov 2018 00:01:05 +0800
Subject: [PATCH] Enable TLS 1.3 draft 23 28 for BoringSSL

---
 ssl/internal.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/ssl/internal.h b/ssl/internal.h
index 087f5fbd8..67ae7a364 100644
--- a/ssl/internal.h
+++ b/ssl/internal.h
@@ -2794,7 +2794,7 @@ struct ssl_ctx_st {
 
   // tls13_variant is the variant of TLS 1.3 we are using for this
   // configuration.
-  tls13_variant_t tls13_variant = tls13_rfc;
+  tls13_variant_t tls13_variant = tls13_all;
 
   bssl::UniquePtr<bssl::SSLCipherPreferenceList> cipher_list;
 
@@ -3123,7 +3123,7 @@ struct ssl_st {
 
   // tls13_variant is the variant of TLS 1.3 we are using for this
   // configuration.
-  tls13_variant_t tls13_variant = tls13_rfc;
+  tls13_variant_t tls13_variant = tls13_all;
 
   // session is the configured session to be offered by the client. This session
   // is immutable.
